### PR TITLE
feat: add EndpointUtil.isAnonymousEndpoint

### LINKED ISSUE #10608: Add a method to EndpointUtil to check if a given URL will end up in an @AnonymousAllowed endpoint
### Description of the bug / feature
Right now you need to manually open up access to public (`@AnonymousAllowed`) endpoints in your Spring Security configuration. It would be beneficial if `EndpointUtil` can check this so that the default Spring Security configuration can handle this case automatically.

### Expected behavior
In addition to the current `isEndpointRequest` in `EndpointUtil` there would be an `isPublicEndpointRequest(HttpServletRequest request)`. At the same time, the `isEndpointRequest` method could be improved to actually check for the endpoint instead of just a prefix.

All needed information about endpoint mapping and access control should already be available as endpoints are based around URL -> endpoint mapping + access control.

### PR DESCRIPTION
**Pull request by Artur-**
Fixes #10608

Based on #10610

### CODE REVIEW

**Comment by haijian-vaadin** [comment]
#10610 has been merged, this one could be rebased?

**Comment by Artur-** [comment]
Done. Is it a problem to change the `spring-web` dependency version?

**Comment by haijian-vaadin** [comment]
> Done. Is it a problem to change the `spring-web` dependency version?

updating the version should be fine. But why not change the `spring.version` property to 5.3.0? so that other spring dependencies would be updated as well?

**Comment by Artur-** [comment]
I can do that but won't this mean dropping support for Spring Boot < 2.4? https://spring.io/blog/2020/11/12/spring-boot-2-4-0-available-now mentions that Spring Boot 2.4 upgrades to Spring 5.3

**Comment by Artur-** [comment]
Fixed. Upgrading the Spring dependency should not be a problem as we are not backporting this change anyway

File: fusion-endpoint/pom.xml [hunk 1]
```diff
@@ -72,7 +71,7 @@
         <dependency>
             <groupId>org.springframework</groupId>
             <artifactId>spring-web</artifactId>
-            <version>${spring.version}</version>
+            <version>${spring.vesion}</version>
```
Comment (haijian-vaadin) [inline fusion-endpoint/pom.xml:74]: vesion->version

File: fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/EndpointUtil.java [hunk 2]
```diff
@@ -29,41 +42,70 @@
...
+                .parse(endpointProperties.getVaadinEndpointPrefix()
+                        + VaadinConnectController.ENDPOINT_METHODS);
+        RequestPath requestPath = ServletRequestPathUtils
+                .parseAndCache(request);
+        if (pathPattern.matches(requestPath)) {
+            PathMatchInfo e = pathPattern.matchAndExtract(requestPath);
+            if (e == null) {
+                return Optional.empty();
+            }
+
```
Comment (haijian-vaadin) [inline fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/EndpointUtil.java:79]: could update the magic variables? e.g.
p->parse
e->mathInfo

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    private void testAnon(String endpointName, String... expectedAnonMethods) {
+        testAnon(endpointName, Arrays.asList(expectedAnonMethods));
+    }
+
+    private void testAnon(String endpointName, List<String> expectedAnonMethods) {
+        for (String endpointMethod : endpointMethods) {
+            testAnon("/connect/" + endpointName + "/" + endpointMethod, expectedAnonMethods.contains(endpointMethod));
+        }
+
+    }
```
Comment (haijian-vaadin) [inline fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java:81]: These two methods could be merged into 1?
Also had to think a while about what `testAnon` means. could maybe change the name to 
`verifyEndpointMethodsAreAnonymousAllowed()`

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        Assert.assertEquals(expected, endpointUtil.isEndpointRequest(request));
+
+        request = Mockito.mock(HttpServletRequest.class);
+        Mockito.when(request.getServletPath()).thenReturn(path);
+        Mockito.when(request.getRequestURI()).thenReturn(path);
+        Assert.assertEquals(expected, endpointUtil.isEndpointRequest(request));
+
+    }
+
+    private void testAnon(String path, boolean expected) {
```
Comment (haijian-vaadin) [inline fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java:104]: had to think even for this `testAnon`, could maybe change that name to
veryEndpointPathIsAnonymous().

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    @Test
+    public void isAnonymousEndpoint() {
+        testAnon("AnonymousAllowedEndpoint", "noannotation", "anonymousallowed");
+        testAnon("DenyAllEndpoint", "anonymousallowed");
+        testAnon("NoAnnotationEndpoint", "anonymousallowed");
+        testAnon("DenyAllEndpoint", "anonymousallowed");
+        testAnon("PermitAllEndpoint", "anonymousallowed");
+        testAnon("RolesAllowedAdminEndpoint", "anonymousallowed");
```
Comment (haijian-vaadin) [inline fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java:68]: Seems it only tests the anonymous endpoints are anonymous, not testing the non-anonymous endpoints are non-anonymous

**Review by haijian-vaadin — CHANGES_REQUESTED** [review]

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java (hunk 5 above)
Comment (Artur-) [reply fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java]: It checks through all endpoints and gets the `expected` value from the strings here

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java (hunk 4 above)
Comment (Artur-) [reply fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java]: Done

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java (hunk 3 above)
Comment (Artur-) [reply fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java:102]: Done

File: fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/EndpointUtil.java (hunk 2 above)
Comment (Artur-) [reply fusion-endpoint/src/main/java/com/vaadin/flow/server/connect/EndpointUtil.java:79]: Done

File: fusion-endpoint/pom.xml (hunk 1 above)
Comment (Artur-) [reply fusion-endpoint/pom.xml]: Done

File: fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java (hunk 5 above)
Comment (haijian-vaadin) [reply fusion-endpoint/src/test/java/com/vaadin/flow/server/connect/EndpointUtilTest.java]: ah, ok, I misread the code.

**Review by haijian-vaadin — APPROVED** [review]

**Review by joheriks — APPROVED** [review]

**Pull request merged**